ddl/job_table.go

```diff
@@ -194,28 +165,41 @@ func hasSysDB(job *model.Job) bool {
 	return false
 }
 
-func (d *ddl) handleUpgradingState(se *sess.Session, job *model.Job) error {
-	if !d.stateSyncer.IsUpgradingState() {
-		if !job.IsPausedBySystem() || hasSysDB(job) {
-			return nil
+func (d *ddl) processJobDuringUpgrade(sess *sess.Session, job *model.Job) (err error, isRunnable bool) {
+	if d.stateSyncer.IsUpgradingState() {
+		if hasSysDB(job) {
+			return err, false
 		}
-		_, err := ResumeJobsBySystem(se.Session(), []int64{job.ID})
-		if err != nil {
-			logutil.BgLogger().Warn("[ddl] resume user DDL by system failed", zap.Stringer("job", job), zap.Error(err))
-			return err
+		var errs []error
+		// During binary upgrade, pause all running DDL jobs
+		errs, err = PauseJobsBySystem(sess.Session(), []int64{job.ID})
+		if len(errs) > 0 {
+			if dbterror.ErrCannotPauseDDLJob.Equal(errs[0]) ||
+				dbterror.ErrPausedDDLJob.Equal(errs[0]) {
+				return nil, false
+			}
+
+			// During upgrade, there are jobs that may not be paused because of `!job.IsRunning`.
+			// Then, we would not run it and just return unless the upgrade finished.
+			return errs[0], false
 		}
-		logutil.BgLogger().Info("[ddl] resume user DDL by system successful", zap.Stringer("job", job))
-		return nil
-	}
-	if job.IsPausing() {
-		return nil
+
+		// no matter how, we would not run DDL jobs during upgrade even not paused.
+		return err, false
 	}
-	if hasSysDB(job) {
-		return nil
+
+	if job.IsPausedBySystem() && !hasSysDB(job) {
+		var errs []error
+		errs, err = ResumeJobsBySystem(sess.Session(), []int64{job.ID})
+		if len(errs) > 0 {
```

Comment:
It seems that `len(errs)` always greater than 0
`errs = make([]error, len(ids))`: https://github.com/pingcap/tidb/pull/43907/files#diff-a185df836b8926c4afe351d4def1b56d1e282c65aabe6677c2fa2c54590cc4b2R1519

Reply:
good one